artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/broker/AMQPMessage.java

```diff
@@ -825,8 +826,14 @@ protected ReadableBuffer createDeliveryCopy(int deliveryCount, DeliveryAnnotatio
       }
 
       writeDeliveryAnnotationsForSendBuffer(result, deliveryAnnotations);
-      // skip existing delivery annotations of the original message
-      duplicate.position(encodedHeaderSize + encodedDeliveryAnnotationsSize);
+
+      if (headerPosition > deliveryAnnotationsPosition && headerPosition != VALUE_NOT_PRESENT && deliveryAnnotationsPosition != VALUE_NOT_PRESENT) {
+         // this is for a case where delivery annotations was swiched wrongly in a previous version
+         duplicate.position(deliveryAnnotationsPosition + encodedDeliveryAnnotationsSize);
```

Comment:
I dont really understand this. Is it trying to _handle_ the delivery-annotations having been written in the wrong place, in front of the header, in an old stored message?

If so, how does setting the position to the end of the delivery annotations do that? Wouldnt that mean it was now positioned at the start of the illegally-positioned header...which we have already just written [if needed] previously a few lines earlier in this method?

Reply:
I tried using headerPosition+headerSize, and didn't work... I ended up thinking it was correct, but now that you raised this I don't know either.. :)


let me review this.. I might need your help

Reply:
basically the test works with what I have now, and it failed with setting the position after the header.

Reply:
I dont understand the issue to the extent you do, but if from the comment in the code you are saying the sections were reversed, then the code as-is simply does not seem correct. If they were something more than reversed, its not clear to me it could be handled. (Its not clear to me that it should be trying at all, its basically a corrupted message).

The tests also passed before, so that may just mean you dont have a focused enough test yet. I probably wouldnt be using Qpid JMS to ultimately check this either, it should be verifying/decoding the payload more closely. Arguably this should be more easily unit testable given its in the message object.

Reply:
@gemmellr the issue manifested by stopping all brokers with messages accumulated. and this is what the test is doing.

It's also mixing versions so that they can be swapped.


I will add more tests but I will keep this test I wrote.

Reply:
in regard to this calculation, the headerPosition was calculated wrongly when the header was switched.... I'm fixing it.